torch/csrc/jit/passes/tensorexpr_fuser.cpp

```diff
@@ -1197,6 +1197,113 @@ class TensorExprFuser {
   }
 #undef REQ
 
+  // Move the given user of `aten::cat` op to its inputs.
+  Node* moveCatAfterUse(
+      Node* cat,
+      Node* user,
+      std::shared_ptr<Graph> subgraph) {
+    // Example IR:
+    //   %1 = ...
+    //   %2 = ...
+    //   %3 = prim::ListConstruct(%1, %2)
+    //   %4 = aten::cat(%3, ...)
+    //   %5 = aten::relu(%4)
+    //   return (%5)
+    //
+    // To be transformed to:
+    //   %1 = ...
+    //   %2 = ...
+    //   %5.1 = aten::relu(%1)
+    //   %5.2 = aten::relu(%2)
+    //   %3 = prim::ListConstruct(%5.1, %5.2)
+    //   %4 = aten::cat(%3, ...)
+    //   return (%4)
+
+    TORCH_INTERNAL_ASSERT(cat->output()->hasUses());
+    TORCH_INTERNAL_ASSERT(cat->output()->uses().size() == 1);
+    TORCH_INTERNAL_ASSERT(cat->input(0)->node()->kind() == prim::ListConstruct);
+    auto cat_list = cat->input(0)->node();
+    auto cat_inputs = cat_list->inputs();
+
+    std::unordered_map<Value*, Value*> new_cat_inputs;
+    for (auto inp : cat_inputs) {
+      auto new_cat_input = subgraph->createClone(
+          user, [&](Value* k) { return (k == cat->output()) ? inp : k; });
+      new_cat_input->output()->setType(inp->type());
+      new_cat_input->insertBefore(cat_list);
+      new_cat_inputs[inp] = new_cat_input->output();
+    }
+    auto new_cat_list = subgraph->createClone(
+        cat_list, [&](Value* k) { return new_cat_inputs[k]; });
+    new_cat_list->insertBefore(cat);
+    auto new_cat = subgraph->createClone(cat, [&](Value* k) {
+      return (k == cat_list->output()) ? new_cat_list->output() : k;
+    });
+    new_cat->insertBefore(cat);
+
+    user->output()->replaceAllUsesWith(new_cat->output());
+    user->destroy();
+
+    TORCH_INTERNAL_ASSERT(!cat->output()->hasUses());
```

Comment:
just fwiw `cat->destroy()` will also check that the output has no uses, but ya maybe it's nice to be explicit here with the assertion before